vault/resource_pki_backend_role.go

```diff
@@ -0,0 +1,380 @@
+			},
+			"email_protection_flag": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Specifies if certificates are flagged for email protection use.",
+			},
+			"key_type": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "rsa",
+				Description: "Specifies the type of key to generate for generated private keys. Currently, rsa and ec are supported.",
+			},
+			"key_bits": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     2048,
+				Description: "Specifies the number of bits to use for the generated keys.",
+			},
+			"key_usage": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: "Policies to be set on tokens issued using this role.",
+			},
+			"use_csr_common_name": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "When used with the CSR signing endpoint, the common name in the CSR will be used instead of taken from the JSON data.",
+			},
+			"use_csr_sans": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "When used with the CSR signing endpoint, the subject alternate names in the CSR will be used instead of taken from the JSON data.",
+			},
+			"ou": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Specifies the OU (OrganizationalUnit) values in the subject field of issued certificates.",
+			},
+			"organization": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Specifies the O (Organization) values in the subject field of issued certificates.",
+			},
+			"generate_lease": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Specifies if certificates issued/signed against this role will have Vault leases attached to them.",
+			},
+			"no_store": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "If set, certificates issued/signed against this role will not be stored in the storage backend.",
+			},
+			"backend": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Unique name of the PKI backend to configure.",
+				ForceNew:    true,
+				Default:     "pki",
+				// standardise on no beginning or trailing slashes
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func pkiBackendRoleWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	role := d.Get("role").(string)
+
+	path := pkiBackendRolePath(backend, role)
+
+	log.Printf("[DEBUG] Writing PKI backend role %q", path)
```

Comment:
Could we move this down so it's right over the actual API call? Just in case future log lines get introduced, or in case of panic, it'd be more helpful to claim we're writing it when we're _actually_ going to be writing it, I think.

Reply:
Absolutely.